2015/presentations/writing-snmp-apps/text.md

```diff
@@ -0,0 +1,470 @@
+boxes did not scale well.
+
+At that time there was no single predominant packet switching technology
+like we see today. Large organizations were busy developing their own
+network stacks each equipped with some form of network management
+facilities. Ultimately, in early 90's, Internet Protocol Suite wins this
+competition practically eliminating all other rivals. For TCP/IP network
+management they offered SNMP, which remains principal technology even
+today.
+
+As a technology, SNMP defines application layer protocol, data model and
+data objects. The protocol, which evolved a great deal since its initial
+introduction in early 80's, serves as a more or less secure, lightweight
+and fault-tolerant communication channel between parties.  SNMP data model
+maps all interesting nuances of host or application internals to named
+variables organized into hierarchy. Concrete collection of variables is
+domain-specific, it is formally defined in MIBs -- files written in a
+special domain-specific language, subset of ASN.1.
+
+Although SNMP designers were trying to kill two birds with one stone,
+offering both information collection and versatile remote configuration
+features, the latter never really enjoyed much popularity among
+implementers. Thus most frequently, SNMP is used for gathering some or all
+variables from hosts or applications being managed. 
+
+With SNMP architecture, managed host or application should have a component
+called SNMP Agent. It acts as an intermediate having access to
+host/application internals and being able to funnel that information over
+SNMP to interested parties in form of SNMP variables. 
+
+The other part of the system is called SNMP Manager, this component is
+always looking for SNMP variables either by querying SNMP Agents or
+listening for notifications they may produce whenever something happens to
+them. 
+
+Library orientation
+-------------------
+
+The PySNMP library is internally structured along the lines of 
+[RFC3411](http://www.ietf.org/rfc/rfc3411.txt). Components that are not
+unique to SNMP are put into stand-alone Python packages to promote
+reusability.
+
+SNMP protocol is defined in terms of ASN.1 data structures, SNMP messages
+travelling the wire are encoded in BER. For those purposes PySNMP relies on
+generic implementation of ASN.1 types and codecs distributed as a dedicated
+Python package under the name of [PyASN1](http://pyasn1.sf.net).
+
+SNMP-level data processing is performed by a collection of SNMP Message 
+Processing ([RFC3412](http://www.ietf.org/rfc/rfc3412.txt)) and Security
+([RFC3414](http://www.ietf.org/rfc/rfc3414.txt)) modules living in 
+*pysnmp.proto...* sub-package. All crypto operations are offloaded to
+the third-party [PyCrypto](https://www.dlitz.net/software/pycrypto/)
+package.
+
+Base classes acting as a wireframe for SMI objects
+([RFC2587](http://www.ietf.org/rfc/rfc2578.txt)) are defined in
+*pysnmp.smi.*... They carry out both of MIB purposes:
```

Comment:
I don't fully understand the sentence starting above, as if something was missing there.